### PR TITLE
solved cut and then paste a checklist turns it into a bullet list #10690

### PR DESCRIPTION
**Pull request by shubhiscoding**
solved cut and then paste a checklist turns it into a bullet list #10690 

[Screencast from 2024-07-09 16-13-43.webm](https://github.com/laurent22/joplin/assets/95687419/44ce159f-c88c-4442-bfb3-07785a81d7b1)

### CODE REVIEW

**Comment by shubhiscoding** [comment]
I have read the CLA Document and I hereby sign the CLA

File: packages/app-desktop/gui/NoteEditor/NoteBody/PlainEditor/PlainEditor.tsx [hunk 1]
```diff
@@ -1,14 +1,76 @@
...
-// Used in safe mode
-
 import * as React from 'react';
 import { ForwardedRef } from 'react';
 import { useEffect, useCallback, useRef, forwardRef, useImperativeHandle } from 'react';
 import { NoteBodyEditorProps, NoteBodyEditorRef } from '../../utils/types';
+const { clipboard } = require('electron');
 
 const PlainEditor = (props: NoteBodyEditorProps, ref: ForwardedRef<NoteBodyEditorRef>) => {
 	const editorRef = useRef<HTMLTextAreaElement>();
```
Comment (personalizedrefrigerator) [inline packages/app-desktop/gui/NoteEditor/NoteBody/PlainEditor/PlainEditor.tsx]: Please include a brief description of how these changes fix the issue.

At present, the `PlainEditor` should only be used in safe mode ([it's initialized here](https://github.com/laurent22/joplin/blob/24954bd0f00f310e9ce233912f175017a84f0e21/packages/app-desktop/gui/NoteEditor/NoteEditor.tsx#L436)). See [TinyMCE.tsx](https://github.com/laurent22/joplin/blob/dev/packages/app-desktop/gui/NoteEditor/NoteBody/TinyMCE/TinyMCE.tsx) for the Rich Text Editor code.

Thank you for working on this!

Comment (shubhiscoding) [reply packages/app-desktop/gui/NoteEditor/NoteBody/PlainEditor/PlainEditor.tsx]: Hii just added the new commit, turns out there was a problem in ```onCut``` function in ```copyHtmlToClipboard(selectedContent);```
when I removed it I found out that on pasting after a cut I was getting one extra square bracket something like this
- [ ] []
test
- [ ] test2

and I found out that the if statement mention below from ```processPastedHtml``` function in resourceHandling.ts was adding that extra square bracket and a line brake so I  explicitly added an extra check such that it doesn't execute when we are pasting a checklist 
```
	if (htmlToMd && mdToHtml) {
		const md = await htmlToMd(MarkupLanguage.Markdown, html, '');
		html = (await mdToHtml(MarkupLanguage.Markdown, md, markupRenderOptions({ bodyOnly: true }))).html;

		// When plugins that add to the end of rendered content are installed, bodyOnly can
		// fail to remove the wrapping paragraph. This works around that issue by removing
		// the wrapping paragraph in more cases. See issue #10061.
		if (!md.trim().includes('\n')) {
			html = removeWrappingParagraphAndTrailingEmptyElements(html);
		}
	}
```

On making these changes I was able to cut and paste properly. The video below is demo of latest changes.

[Screencast from 2024-07-12 11-19-55.webm](https://github.com/user-attachments/assets/ef2726ee-0cbb-41c2-a8d5-796b3604fc8c)


While raising the PR earlier the cut past worked on video because I think I may have mistakenly perform copy instead of cut😅

File: packages/app-desktop/gui/NoteEditor/NoteBody/TinyMCE/TinyMCE.tsx [hunk 2]
```diff
@@ -1229,8 +1229,8 @@ const TinyMCE = (props: NoteBodyEditorProps, ref: any) => {
 
 		// eslint-disable-next-line @typescript-eslint/no-explicit-any -- Old code before rule was applied
 		async function onCut(event: any) {
-			const selectedContent = editor.selection.getContent();
-			copyHtmlToClipboard(selectedContent);
```
Comment (roman-r-m) [inline packages/app-desktop/gui/NoteEditor/NoteBody/TinyMCE/TinyMCE.tsx:1233]: so you removed the line where copying to clipboard took place, how is that even supposed to work?

File: packages/app-desktop/gui/NoteEditor/utils/resourceHandling.ts [hunk 3]
```diff
@@ -212,12 +212,11 @@ export async function processPastedHtml(html: string, htmlToMd: HtmlToMarkdownHa
...
 
 	html = await processImagesInPastedHtml(html);
-
 	// TinyMCE can accept any type of HTML, including HTML that may not be preserved once saved as
 	// Markdown. For example the content may have a dark background which would be supported by
 	// TinyMCE, but lost once the note is saved. So here we convert the HTML to Markdown then back
-	// to HTML to ensure that the content we paste will be handled correctly by the app.
-	if (htmlToMd && mdToHtml) {
+	// // to HTML to ensure that the content we paste will be handled correctly by the app.
+	if (!html.includes('joplin-checklist') && htmlToMd && mdToHtml) {
```
Comment (roman-r-m) [inline packages/app-desktop/gui/NoteEditor/utils/resourceHandling.ts]: what if the html content is something like
```
<p>joplin-checklist</p>
````

File: packages/app-desktop/gui/NoteEditor/NoteBody/PlainEditor/PlainEditor.tsx [hunk 4]
```diff
@@ -1,14 +1,76 @@
-
```
Comment (roman-r-m) [inline packages/app-desktop/gui/NoteEditor/NoteBody/PlainEditor/PlainEditor.tsx]: now that we have established that this wasn't the right file to change, please revert the changes in this file

Comment (shubhiscoding) [reply packages/app-desktop/gui/NoteEditor/NoteBody/PlainEditor/PlainEditor.tsx]: reverted the change in latest commit

File: packages/app-desktop/gui/NoteEditor/NoteBody/TinyMCE/TinyMCE.tsx (hunk 2 above)
Comment (shubhiscoding) [reply packages/app-desktop/gui/NoteEditor/NoteBody/TinyMCE/TinyMCE.tsx:1233]: > so you removed the line where copying to clipboard took place, how is that even supposed to work?

When I copy paste an entire checklist I saw in the logs that the pasted HTML had a class name joplin-checklist, and when I saw the logs of cut event I didn't saw the class given to the checklist, so I removed  ```copyHtmlToClipboard(selectedContent);``` to see if there's any default/global cut event working and I saw something like this when I cut pasted after removing ```copyHtmlToClipboard(selectedContent);```
 - [ ] []
- [ ] test
- [ ] test2

There was an extra square bracket added to the selected content, but I wasn't getting bullet markers instead on checkbox so, the cut was working fine, I solved this extra bracket problem in changes of resourceHandling.ts

File: packages/app-desktop/gui/NoteEditor/utils/resourceHandling.ts (hunk 3 above)
Comment (shubhiscoding) [reply packages/app-desktop/gui/NoteEditor/utils/resourceHandling.ts]: changed this to -> ``` !html.includes('class="joplin-checklist"') ```in latest commit, so that it's only excluded when there's checkbox

**Comment by roman-r-m** [comment]
This is still not good.
1. There's now an error logged in console when cutting
2. Pasting code block produces html.

**Comment by laurent22** [comment]
@shubhiscoding, please confirm if you are able to address Roman's concerns. I agree it cannot be merged as it is.
